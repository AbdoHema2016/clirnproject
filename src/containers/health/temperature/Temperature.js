import React from 'react';
import {View, SafeAreaView, Platform, Text} from 'react-native';
import Style from './style';
import {connect} from 'react-redux';
import CLabel from '../../../components/cLabel/index';
import CButton from '../../../components/cButton/index';
import CButtonWithImage from '../../../components/cButtonWithImage/index';
import navigationService from '../../../Navigation/NavigationService';
import CPicker from '../../../components/cPicker';
import {
  setTempTypeAction,
  setTempIntValAction,
  setTempDecValAction,
  sendTempAction,
} from './redux/actions';
import {signUpStepAction} from '../../onBoarding/personalDetails/redux/actions';
import dayjs from 'dayjs';
import AsyncStorage, {AsyncConstants} from '../../../utilities/AsyncStorage';
import CNavigationBackButton from '../../../components/cNavigationBackButton';
import {
  DATEFORMATS,
  signUpSteps,
  LOCAL_PATH,
  temperatureDefaultValues,
  showErrorMessage,
  temperatureMaxRangeLimit,
  temperatureType,
  testIds,
  analyticsIds,
} from '../../../utilities';
import {translate} from '../../../Localization';

import {logAnalyticsEvent} from '../../../utilities/Firebase';

class Temperature extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      showPicker: false,
      intOrDec: '',
      intVal: 0,
      decVal: 0,
    };
    this.nextextInputRef = {};
    navigationService.navigation = this.props.navigation;
  }

  componentDidMount() {
    const {celsius, fahrenheit} = temperatureDefaultValues;
    this.props.setTempIntVal(
      this.props.temperature.tempType === '°F' ? fahrenheit : celsius,
    );
    this.props.setTempDecVal('0');
    this.navigationOptions();
  }

  navigationOptions = () => {
    this.props.navigation.setOptions({
      title: '',
      headerTransparent: true,
      headerLeft: () => (
        <CNavigationBackButton
          title={translate('BACK.BACK_BUTTON_TITLE')}
          backButtonAction={this.goBack}
        />
      ),
    });
  };

  goBack = () => {
    this.props.navigation.navigate('Feeling');
  };

  tempScaleType = (val) => {
    const {celsius, fahrenheit} = temperatureDefaultValues;
    let intVal = val === '°F' ? fahrenheit : celsius;
    this.props.setTempType(val, intVal, '0');
  };

  hidePicker = () => this.setState({showPicker: false});
  showPicker = (val) => {
    this.setState({
      showPicker: val,
    });
  };

  selectTemperatureValue = (type, val) => {
    this.showPicker(true);
    if (val === null) {
      this.setState({
        intOrDec: type,
      });
    } else {
      if (type === 'INT') {
        this.props.setTempIntVal(val);
      } else {
        this.props.setTempDecVal(val);
      }
    }
  };

  onSkipPressed = () => {
    this.props.signupStep('7');
    AsyncStorage.setItemInStorage(AsyncConstants.STEP, signUpSteps.HealthTest);
    this.props.navigation.navigate('HealthTest');
    logAnalyticsEvent(analyticsIds.onboard_temp_skipped);
  };

  sendTemperatureData = () => {
    const {
      temperature: {tempType, tempIntValue, tempDecValue},
      access_token: token,
      statId,
      sendTempData,
      userID,
    } = this.props;
    const celsiusDecimalAndFractionSplitArray = temperatureMaxRangeLimit.celsius.split(
      '.',
    );

    const celsiusMaxDecimalPlace = celsiusDecimalAndFractionSplitArray[0];
    const celsiusMaxFractionPlace = celsiusDecimalAndFractionSplitArray[1];

    const fahrenheitDecimalAndFractionSplitArray = temperatureMaxRangeLimit.fahrenheit.split(
      '.',
    );

    const fahrenheitMaxDecimalPlace = fahrenheitDecimalAndFractionSplitArray[0];
    const fahrenheitMaxFractionPlace =
      fahrenheitDecimalAndFractionSplitArray[1];

    if (
      tempType === temperatureType.celsiusWithDegree &&
      tempIntValue === celsiusMaxDecimalPlace &&
      tempDecValue > celsiusMaxFractionPlace
    ) {
      const errorMessage =
        translate('STRINGS.MAX_TEMP_LIMIT') +
        temperatureMaxRangeLimit.celsius +
        temperatureType.celsiusWithDegree;

      return showErrorMessage(errorMessage);
    }

    if (
      tempType === temperatureType.fahrenheitWithDegree &&
      tempIntValue === fahrenheitMaxDecimalPlace &&
      tempDecValue > fahrenheitMaxFractionPlace
    ) {
      const errorMessage =
        translate('STRINGS.MAX_TEMP_LIMIT') +
        temperatureMaxRangeLimit.fahrenheit +
        temperatureType.fahrenheitWithDegree;

      return showErrorMessage(errorMessage);
    }
    let data = {
      date_of_reading: dayjs(new Date()).format(DATEFORMATS.TEMPERATURE_DATA),
      temperature: tempIntValue + '.' + tempDecValue,
      preferred_measurement: tempType === '°C' ? '1' : '2',
      user_id: userID,
      step: 6,
    };

    sendTempData(data, token, statId);
    logAnalyticsEvent(analyticsIds.onboard_temp_added);
  };

  render() {
    const {
      temperature: {
        tempType,
        tempIntValue,
        error,
        celciusDegrees,
        fahrenheit,
        tempDecValue,
        decimalPart,
        loading,
      },
    } = this.props;
    const {intOrDec} = this.state;

    return (
      <SafeAreaView style={Style.container}>
        <View style={Style.lablesContainer}>
          <CLabel style={Style.infoStageLabel} text={'6/7'} />
          <CLabel
            style={Style.appTitle}
            text={translate('STRINGS.TEMPERATURE_QUERY')}
          />
          <CLabel
            style={Style.chooseScale}
            text={translate('STRINGS.TEMPERATURE_SCALE')}
          />
        </View>

        <View style={Style.tempScaleType}>
          <CButton
            testID={testIds.selectFeh}
            accessibilityLabel={testIds.selectFeh}
            onPress={() => this.tempScaleType('°F')}
            text={'°F'}
            textStyle={Style.blackText}
            buttonContainerStyle={
              tempType === '°F'
                ? Style.selectedTemperatureScaleTypeButtonContainer
                : Style.unSelectedTemperatureScaleTypeButtonContainer
            }
            buttonCustomStyle={Style.temperatureScaleTypeButton}
          />
          <CButton
            testID={testIds.selectCel}
            accessibilityLabel={testIds.selectCel}
            onPress={() => this.tempScaleType('°C')}
            text={'°C'}
            textStyle={Style.blackText}
            buttonContainerStyle={
              tempType === '°C'
                ? Style.selectedTemperatureScaleTypeButtonContainer
                : Style.unSelectedTemperatureScaleTypeButtonContainer
            }
            buttonCustomStyle={Style.temperatureScaleTypeButton}
          />
        </View>

        <View style={Style.tempScaleType}>
          {Platform.OS === 'ios' ? (
            <CButtonWithImage
              testID={testIds.SelInt}
              accessibilityLabel={testIds.SelInt}
              onPress={() => this.selectTemperatureValue('INT', null)}
              text={tempIntValue}
              textStyle={Style.blackText}
              imagePath={LOCAL_PATH.DROP_DOWN_ICON}
              buttonContainerStyle={
                Style.selTemperatureIntegerValueButtonContainer
              }
              buttonCustomStyle={Style.temperatureIntegerValueButton}
            />
          ) : (
            <CPicker
              testID={testIds.IntPicker}
              accessibilityLabel={testIds.IntPicker}
              androidPickerContainer={
                Style.selTemperatureIntegerValueButtonContainer
              }
              androidPicker={Style.temperatureIntegerValueButton}
              selectedValue={tempIntValue}
              setTitle={(val) => this.selectTemperatureValue('INT', val)}
              showDatePicker={this.showPicker}
              error={error}
              pickerItems={tempType === '°C' ? celciusDegrees : fahrenheit}
              imagePath={LOCAL_PATH.DROP_DOWN_ICON}
            />
          )}

          <CLabel style={Style.dotLabel} text={'.'} />
          {Platform.OS === 'ios' ? (
            <CButtonWithImage
              testID={testIds.SelDec}
              accessibilityLabel={testIds.SelDec}
              onPress={() => this.selectTemperatureValue('DEC', null)}
              text={tempDecValue}
              textStyle={Style.blackText}
              imagePath={LOCAL_PATH.DROP_DOWN_ICON}
              buttonContainerStyle={
                Style.selTemperatureIntegerValueButtonContainer
              }
              buttonCustomStyle={Style.temperatureIntegerValueButton}
            />
          ) : (
            <CPicker
              testID={testIds.DecPicker}
              accessibilityLabel={testIds.DecPicker}
              androidPicker={Style.temperatureIntegerValueButton}
              androidPickerContainer={
                Style.selTemperatureIntegerValueButtonContainer
              }
              selectedValue={tempDecValue}
              setTitle={(val) => this.selectTemperatureValue('DEC', val)}
              showDatePicker={this.showPicker}
              error={error}
              pickerItems={decimalPart}
              imagePath={LOCAL_PATH.DROP_DOWN_ICON}
            />
          )}

          <CLabel style={Style.tempScaleLabel} text={tempType} />
        </View>

        <View style={Style.buttonsContainer}>
          <CButton
            testID={testIds.continue}
            accessibilityLabel={testIds.continue}
            text={translate('LOGIN_SCREEN_STRINGS.CONTINUE_LABEL')}
            textStyle={Style.whiteText}
            onPress={() => this.sendTemperatureData()}
            buttonContainerStyle={Style.abledGreenButton}
            loading={loading}
          />

          <Text
            accessible={true}
            testID={testIds.skip}
            accessibilityLabel={testIds.skip}
            style={Style.skip}
            onPress={() => this.onSkipPressed()}>
            {translate('STRINGS.SKIP')}
          </Text>
        </View>
        {this.state.showPicker && Platform.OS === 'ios' ? (
          <CPicker
            testID={testIds.DecPicker}
            accessibilityLabel={testIds.DecPicker}
            selectedValue={intOrDec === 'DEC' ? tempDecValue : tempIntValue}
            setTitle={(val) => this.selectTemperatureValue(intOrDec, val)}
            showDatePicker={this.hidePicker}
            pickerItems={
              tempType === '°C'
                ? intOrDec === 'INT'
                  ? celciusDegrees
                  : decimalPart
                : intOrDec === 'INT'
                ? fahrenheit
                : decimalPart
            }
          />
        ) : null}
      </SafeAreaView>
    );
  }
}

const mapStateToProps = ({
  signIn: {userID},
  selectTempType,
  feeling,
  personalDetails,
  userProfile: {
    userInfo: {email},
  },
}) => {
  return {
    email,
    temperature: selectTempType,
    statId: feeling.id,
    access_token: personalDetails.access_token,
    userID,
  };
};

const mapDispatchToProps = {
  setTempType: setTempTypeAction,
  setTempIntVal: setTempIntValAction,
  setTempDecVal: setTempDecValAction,
  sendTempData: sendTempAction,
  signupStep: signUpStepAction,
};

export default connect(mapStateToProps, mapDispatchToProps)(Temperature);
