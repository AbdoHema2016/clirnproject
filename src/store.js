import {createStore, applyMiddleware} from 'redux';
import createSagaMiddleware from 'redux-saga';
import {createLogger} from 'redux-logger';
import rootReducer from './reducers';
import {rootSaga} from './sagas';
const sagaMiddleware = createSagaMiddleware();
const isDevelopmentEnv =
  process && process.env && process.env.NODE_ENV === 'development';
const isProductionEnv =
  process && process.env && process.env.NODE_ENV === 'production';
const logStoreInfo = false;
const middlewares = [sagaMiddleware];
function actionLogger({getState}) {
  return (next) => (action) => {
    let allowed = [
      'jumio',
      'notificationsReducer',
      'otp',
      'companies',
      'signIn',
      'updateContactDetailsReducer',
      'userProfile',
      'vaccinations',
    ];
    let state = getState();
    let loggedStates = Object.keys(state)
      .filter((key) => allowed.includes(key))
      .reduce((obj, key) => {
        return {
          ...obj,
          [key]: state[key],
        };
      }, {});
    console.info({actionType: action.type, loggedStates});
    const returnValue = next(action);
    return returnValue;
  };
}
if (isDevelopmentEnv && logStoreInfo) {
  middlewares.push(createLogger());
}
if (isProductionEnv) {
  middlewares.push(actionLogger);
}

const store = createStore(rootReducer, applyMiddleware(...middlewares));
sagaMiddleware.run(rootSaga);
export default store;
