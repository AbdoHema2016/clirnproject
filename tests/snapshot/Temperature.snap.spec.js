import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';
import store from '../../src/store';
import Temperature from '../../src/containers/health/temperature/Temperature';

jest.mock('@react-navigation/native', () => ({
  withNavigationFocus: (component) => component,
}));
jest.useFakeTimers();

describe('Temperature screen', () => {
  const props = {
    navigation: {
      setOptions: jest.fn(),
    },
  };
  it('Snapshot Temperature', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Temperature {...props} />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
