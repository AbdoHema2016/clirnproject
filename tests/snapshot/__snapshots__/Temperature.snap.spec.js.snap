// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Temperature screen Snapshot Temperature 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#F7F7F7",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "width": 724,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#8D8C91",
            "fontFamily": "SofiaPro-Medium",
            "fontSize": 16,
            "marginTop": 160.07999999999998,
            "textAlign": "left",
          },
        ]
      }
    >
      6/7
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#2B342C",
            "fontFamily": "SofiaPro-Medium",
            "fontSize": 24,
            "fontWeight": "600",
          },
        ]
      }
    >
      [missing "en.STRINGS.TEMPERATURE_QUERY" translation]
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#8D8C91",
            "fontFamily": "SofiaPro-Medium",
            "fontSize": 18,
            "marginTop": 26,
          },
        ]
      }
    >
      [missing "en.STRINGS.TEMPERATURE_SCALE" translation]
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "marginTop": 104.052,
        "width": 750,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 4,
            "height": 52,
            "justifyContent": "center",
            "shadowColor": "#DDDDDD",
            "shadowOffset": Object {
              "height": 8,
              "width": 0,
            },
            "shadowOpacity": 0.75,
            "shadowRadius": 4,
          },
          Object {
            "backgroundColor": "#FFFFFF",
            "height": 49,
            "marginLeft": 13,
            "width": 48,
          },
        ]
      }
    >
      <View
        accessibilityLabel="SelectFeh"
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderRadius": 4,
            "height": "100%",
            "justifyContent": "center",
            "opacity": 1,
            "width": "100%",
          }
        }
        testID="SelectFeh"
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "fontFamily": "SofiaProRegular",
                "fontSize": 17,
                "textAlign": "center",
              },
              Object {
                "color": "#2B342C",
                "left": 16,
                "position": "absolute",
              },
            ]
          }
        >
          °F
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 4,
            "height": 52,
            "justifyContent": "center",
            "shadowColor": "#DDDDDD",
            "shadowOffset": Object {
              "height": 8,
              "width": 0,
            },
            "shadowOpacity": 0.75,
            "shadowRadius": 4,
          },
          Object {
            "backgroundColor": "#FFFFFF",
            "borderColor": "#02BF8E",
            "borderWidth": 1.5,
            "height": 49,
            "marginLeft": 13,
            "width": 48,
          },
        ]
      }
    >
      <View
        accessibilityLabel="SelectCel"
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderRadius": 4,
            "height": "100%",
            "justifyContent": "center",
            "opacity": 1,
            "width": "100%",
          }
        }
        testID="SelectCel"
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "fontFamily": "SofiaProRegular",
                "fontSize": 17,
                "textAlign": "center",
              },
              Object {
                "color": "#2B342C",
                "left": 16,
                "position": "absolute",
              },
            ]
          }
        >
          °C
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "marginTop": 104.052,
        "width": 750,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderRadius": 4,
            "flexDirection": "row",
            "height": 48,
            "marginTop": 4,
          },
          Object {
            "borderColor": "#02BF8E",
            "borderWidth": 1.5,
            "height": 48,
            "marginLeft": 13,
            "width": 116,
          },
          null,
        ]
      }
    >
      <View>
        <View>
          34
        </View>
        <View>
          35
        </View>
        <View>
          36
        </View>
        <View>
          37
        </View>
        <View>
          38
        </View>
        <View>
          39
        </View>
        <View>
          40
        </View>
        <View>
          41
        </View>
        <View>
          42
        </View>
      </View>
      <Image
        source={
          Object {
            "ReactComponent": [Function],
          }
        }
        style={
          Object {
            "height": 9,
            "position": "absolute",
            "right": 18,
            "width": 18,
          }
        }
      />
    </View>
    <Text
      style={
        Array [
          Object {
            "marginLeft": 12,
            "textAlign": "center",
            "width": 4,
          },
        ]
      }
    >
      .
    </Text>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderRadius": 4,
            "flexDirection": "row",
            "height": 48,
            "marginTop": 4,
          },
          Object {
            "borderColor": "#02BF8E",
            "borderWidth": 1.5,
            "height": 48,
            "marginLeft": 13,
            "width": 116,
          },
          null,
        ]
      }
    >
      <View>
        <View>
          0
        </View>
        <View>
          1
        </View>
        <View>
          2
        </View>
        <View>
          3
        </View>
        <View>
          4
        </View>
        <View>
          5
        </View>
        <View>
          6
        </View>
        <View>
          7
        </View>
        <View>
          8
        </View>
        <View>
          9
        </View>
      </View>
      <Image
        source={
          Object {
            "ReactComponent": [Function],
          }
        }
        style={
          Object {
            "height": 9,
            "position": "absolute",
            "right": 18,
            "width": 18,
          }
        }
      />
    </View>
    <Text
      style={
        Array [
          Object {
            "fontFamily": "SofiaProRegular",
            "fontSize": 18,
            "marginLeft": 12,
            "marginTop": 12,
            "textAlign": "center",
            "width": 32,
          },
        ]
      }
    >
      °C
    </Text>
  </View>
  <View
    style={
      Object {
        "bottom": 158.74599999999998,
        "height": 50,
        "justifyContent": "space-between",
        "position": "absolute",
        "width": 722,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 4,
            "height": 52,
            "justifyContent": "center",
            "shadowColor": "#DDDDDD",
            "shadowOffset": Object {
              "height": 8,
              "width": 0,
            },
            "shadowOpacity": 0.75,
            "shadowRadius": 4,
          },
          Object {
            "backgroundColor": "#02BF8E",
          },
        ]
      }
    >
      <View
        accessibilityLabel="Continue"
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": undefined,
            "borderRadius": 4,
            "height": "100%",
            "justifyContent": "center",
            "opacity": 1,
            "width": 718,
          }
        }
        testID="Continue"
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "fontFamily": "SofiaProRegular",
                "fontSize": 17,
                "textAlign": "center",
              },
              Object {
                "color": "#FFFFFF",
              },
            ]
          }
        >
          [missing "en.LOGIN_SCREEN_STRINGS.CONTINUE_LABEL" translation]
        </Text>
      </View>
    </View>
    <Text
      accessibilityLabel="Skip"
      accessible={true}
      onPress={[Function]}
      style={
        Object {
          "alignSelf": "center",
          "color": "#2B342C",
          "fontFamily": "SofiaProRegular",
          "fontSize": 16,
          "marginTop": 5,
          "paddingVertical": 15,
          "textAlign": "center",
          "width": 694,
        }
      }
      testID="Skip"
    >
      [missing "en.STRINGS.SKIP" translation]
    </Text>
  </View>
</View>
`;
